crates/bevy_ui/src/layout/mod.rs

```diff
@@ -168,68 +165,83 @@ without UI components as a child of an entity with UI components, results may be
+                .entry((camera_id, entity))
+                .or_insert_with(|| self.taffy.new_leaf(viewport_style.clone()).unwrap());
+
+            if let Some(previous_parent) = self.taffy.parent(user_root_node) {
+                // remove the root node from the previous implicit node's children
+                self.taffy
+                    .remove_child(previous_parent, user_root_node)
+                    .unwrap();
+            }
 
-                    RootNodePair {
-                        implicit_viewport_node: viewport_node,
-                        user_root_node: node,
-                    }
-                });
-            new_roots.push(root_node);
+            self.taffy
+                .add_child(implicit_viewport_node, user_root_node)
+                .unwrap();
         }
-
-        self.camera_roots.insert(camera_id, new_roots);
     }
 
     /// Compute the layout for each window entity's corresponding root node in the layout.
     pub fn compute_camera_layout(&mut self, camera: Entity, render_target_resolution: UVec2) {
-        let Some(camera_root_nodes) = self.camera_roots.get(&camera) else {
-            return;
-        };
+        for (&(camera_entity, _), &implicit_viewport_node) in
+            self.camera_root_to_viewport_taffy.iter()
+        {
+            if camera_entity != camera {
+                continue;
+            }
+
+            let available_space = taffy::geometry::Size {
+                width: taffy::style::AvailableSpace::Definite(render_target_resolution.x as f32),
+                height: taffy::style::AvailableSpace::Definite(render_target_resolution.y as f32),
+            };
 
-        let available_space = taffy::geometry::Size {
-            width: taffy::style::AvailableSpace::Definite(render_target_resolution.x as f32),
-            height: taffy::style::AvailableSpace::Definite(render_target_resolution.y as f32),
-        };
-        for root_nodes in camera_root_nodes {
             self.taffy
-                .compute_layout(root_nodes.implicit_viewport_node, available_space)
+                .compute_layout(implicit_viewport_node, available_space)
                 .unwrap();
         }
     }
 
-    /// Removes each camera entity from the internal map and then removes their associated node from taffy
+    /// Removes specified camera entities by disassociating them from their associated `implicit_viewport_node`
+    /// in the internal map, and subsequently removes the `implicit_viewport_node`
+    /// from the `taffy` layout engine for each.
     pub fn remove_camera_entities(&mut self, entities: impl IntoIterator<Item = Entity>) {
         for entity in entities {
-            if let Some(camera_root_node_map) = self.camera_entity_to_taffy.remove(&entity) {
-                for (_, node) in camera_root_node_map.iter() {
-                    self.taffy.remove(*node).unwrap();
-                }
-            }
+            self.camera_root_to_viewport_taffy.retain(
```

Comment:
When I originally wrote this earlier, I hadn't realized I was retaining inside a loop. I will need to circle back to one.